README.md
[![Codeship](https://img.shields.io/codeship/3b024750-eb65-0132-bd94-5ecc4fff82a0.svg)](https://codeship.com/projects/83402)
[![npm](https://img.shields.io/npm/v/ember-cli-cordova-shims.svg)](https://www.npmjs.com/package/ember-cli-cordova-shims)
[![npm](https://img.shields.io/npm/dm/ember-cli-cordova-shims.svg)](https://www.npmjs.com/package/ember-cli-cordova-shims)
[![npm](https://img.shields.io/npm/l/ember-cli-cordova-shims.svg)](https://www.npmjs.com/package/ember-cli-cordova-shims)

# Ember CLI Cordova Shims

This Ember addon makes it simple to use a range of different Cordova plugins.

## Supported features

* [Push Notifications](#push-notifications)
* [Dialogs](#dialogs)
* [Network Activity Indicator](#network-activity-indicator)

### Push Notifications

Dependencies:

* `cordova plugin add org.apache.cordova.device`
* `cordova plugin add https://github.com/phonegap-build/PushPlugin.git`

Automatic behavior:

* Badge count will be set on the application on incoming notification
* Sounds will play on incoming notification (if set in the payload)

Events:

* `alert` - passes the message in the notification
* `badge` - passes a badge count with every emitted event
* `sound` - passes the filename of the sound to play

Public Methods:

* `register` - Prompts the user to allow notifications to this device
* `unregister` - Revokes notification subscription for this device

Usage:

```javascript
// app/initializers/pushplugin.js
import NotificationsService from 'ember-cli-cordova-shims/services/notifications';

export function initialize(container, application) {
  let service = NotificationsService.create({
    gcmSenderId: '{INSERT-KEY-FROM-GCM-HERE}',
    gcmTimeout: 10000 // Timeout for GCM in ms. Default: 15000
  });
  application.register('service:notifications', service, {
    instantiate: false
  });
  application.inject('route', 'notifications', 'service:notifications');
}

export default {
  name: 'notifications-service',
  initialize: initialize
};
```

```javascript
// app/routes/application.js
import Em from 'ember';

export default Em.Route.extend({

  badge: 0,

  listenForNotifications: function(){
    this.notifications.on('alert', function(message){
      alert(message); // or do whatever you want
    });
    this.notifications.on('badge', (badgeCount) => {
      this.set('badge', badgeCount); // or do whatever you want
    });
  }.on('init'),

  actions: {
    subscribeForNotifications: function(){
      let store = this.store;

      this.notifications.register().then(function(data){
        let network = data.network,
              token = data.token;
        return store.createRecord('device', {
          network: network,
          token: token
        }).save();
      }).then(function(deviceModelFromStore){
        Em.Logger.info('Successfully registered for notifications');
      }).catch(function(error){
        Em.Logger.error('Something went wrong registering for notifications', error);
      });
    },

    unsubscribeNotifications: function(){
      this.notifications.unregister().then(function(){
        Em.Logger.info('Successfully unregistered the device');
      }).catch(function(error){
        Em.Logger.error('Something went wrong while unregistering', error);
      });
    }
  }
});
```

Caveats:

* It's not possible to know what device what unregistered when calling `unregister`
* PushPlugin has a weird setup for registering devices with GCM. There is a timeout
of 15 seconds when doing a registration with GCM. If GCM does not send a `registered`
event within this timeframe, the device is not registered. The timeout value may be
overridden by passing `gcmTimeout` when instantiating the service.

### Dialogs

Dependencies:

* `cordova plugin add https://github.com/apache/cordova-plugin-dialogs.git`

Public Methods:

* `alert` - Show a message to the user
* `confirm` - Ask for some confirmation from the user
* `prompt` - Ask for some information from the user

Usage:

```javascript
// app/initializers/dialogs-service.js
import DialogsService from 'ember-cli-cordova-shims/services/dialogs';

export function initialize(container, application) {
  application.register('service:dialogs', DialogsService);
  application.inject('route', 'dialogs', 'service:dialogs');
}

export default {
  name: 'dialogs-service',
  initialize: initialize
};
```

```javascript
// app/routes/application.js
import Em from 'ember';

export default Em.Route.extend({

  actions: {
    something: function(){
      this.dialogs.alert({
        title: 'Hey',
        message: 'This just happened',
        button: 'OK'
      }).then(function(){
        // Do something
      });
    }
  }
});
```

### Network Activity Indicator

Dependencies:

* `cordova plugin add com.wearecocoon.cordova.plugin.networkactivity`

Usage:

```javascript
// app/adaptesr/application.js
import NetworkActivityAdapterMixin from 'ember-cli-cordova-shims/mixins/network-activity-adapter';

export default DS.RESTAdapter.extend(NetworkActivityAdapterMixin, {
  // Your implementation goes here
});
```

## Installation

* `git clone` this repository
* `npm install`
* `bower install`

## Running

* `ember server`
* Visit your app at http://localhost:4200.

## Running Tests

* `ember test`
* `ember test --server`

## Documentation

The modules are documented with yuidoc.

* `npm install yuidocjs -g`
* `yuidoc -c yuidoc.json`

## Building

* `ember build`

For more information on using ember-cli, visit [http://www.ember-cli.com/](http://www.ember-cli.com/).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [Unreleased][unreleased]


## [0.3.0] - 2015-06-02

### Added
- NetworkActivityAdapterMixin

### Fixed
- Fix module name references in docs and README


## [0.2.1] - 2015-06-01

### Fixed
- Fix issue with missing plugin helper


## [0.2.0] - 2015-06-01

### Added
- Dialogs service to show alert, confirm and prompt


## [0.1.0] - 2015-05-29

### Added
- Notifications service to subscribe/unsubscribe and handle push notifications


[unreleased]: https://github.com/Skalar/ember-cli-cordova-shims/compare/v0.3.0...HEAD
[0.3.0]: https://github.com/Skalar/ember-cli-cordova-shims/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/Skalar/ember-cli-cordova-shims/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/Skalar/ember-cli-cordova-shims/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/Skalar/ember-cli-cordova-shims/compare/10558788...v0.1.0


addon/services/notifications.js
import Em from 'ember';
import missingPlugin from '../utils/missing-plugin';

/**
  Service that allows for registration/unregistration and handling of push
  notifications received from GCM or APNs.

  Configuration and setup of the service is done in an initializer:

  ```javascript
  import NotificationsService from 'ember-cli-cordova-shims/services/notifications';

  export function initialize(container, application) {
    let service = NotificationsService.create({
      gcmSenderId: '{INSERT-KEY-FROM-GCM-HERE}',
      gcmTimeout: 10000 // Timeout for GCM in ms. Default: 15000
    });
    application.register('service:notifications', service, {
      instantiate: false
    });
    application.inject('route', 'notifications', 'service:notifications');
  }

  export default {
    name: 'notifications-service',
    initialize: initialize
  };
  ```

  @class Notifications
  @namespace CordovaShims.Services
  @module ember-cli-cordova-shims/services/notifications
  @extends Ember.Service
  @since 0.1.0
 */
export default Em.Service.extend(Em.Evented, {
  /**
   @event alert
   @param {String} message The message received in the payload
  */

  /**
   @event badge
   @param {Number} count Badge count recieved in the payload
  */

  /**
   @event sound
   @param {String} filename Filename of the sound that should be played
  */

  /**
   Please do not attach event listeners to this event. Use the `register` method
   instead to handle device registration.

   @event registered
   @private
   @param {Object} data An object with a `network` and a `token`. `network` will
                        be either 'apns' or 'gcm'.
  */

  /**
    Project ID from Google API Console. Must be set if you want to register
    devices with GCM.

    @property gcmSenderId
    @type String
  */
  gcmSenderId: null,

  /**
    Timeout for GCM registration. In ms.

    @property gcmTimeout
    @type Number
    @default 15000
  */
  gcmTimeout: 15000,

  init: function(){
    this._subscribeForEcb();
    this.on('badge', this, this._setBadgeCount);
    this.on('sound', this, this._playSound);
    if (!window.device) {
      missingPlugin('Device', 'org.apache.cordova.device');
    }
    if (!window.plugins || !window.plugins.pushNotification) {
      missingPlugin('PushPlugin', 'https://github.com/phonegap-build/PushPlugin.git');
    }
  },

  /**
    Registers the device for push notifications.

    Example:

    ```javascript
    // app/routes/application.js
    import Em from 'ember';

    export default Em.Route.extend({
      actions: {
        subscribeForNotifications: function(){
          let store = this.store;

          this.notifications.register().then(function(data){
            let network = data.network,
                  token = data.token;
            return store.createRecord('device', {
              network: network,
              token: token
            }).save();
          }).then(function(deviceModelFromStore){
            Em.Logger.info('Successfully registered for notifications');
          }).catch(function(error){
            Em.Logger.error('Something went wrong registering for notifications', error);
          });
        }
      }
    });
    ```

    @method register
    @return {RSVP.Promise} A promise that will resolve with an object containing
                           the `network` (apns or gcm) and the `token` for the
                           device.
  */
  register: function(){
    return new Em.RSVP.Promise((resolve, reject) => {
      const pushNotification = window.plugins.pushNotification;
      let success = resolve;

      this.one('_registration', resolve);
      switch (window.device.platform.toLowerCase()) {
        case 'android':
        case 'amazon-fireos':
          let gcmSenderId = this.get('gcmSenderId'),
              gcmTimeout  = this.get('gcmTimeout');

          Em.assert('Attempted to register device for notifications with GCM Sender ID', gcmSenderId);

          Em.Logger.info('Attempting registration with GCM');
          success = function(){
            // Wait for registration event for up to 15 seconds
            setTimeout(function(){
              reject(new Error('Timed out while registering for GCM'));
            }, gcmTimeout);
          };
          pushNotification.register(success, reject, {
            senderId: gcmSenderId,
            ecb: "onNotificationGCM"
          });
          break;
        case 'ios':
          Em.Logger.info('Attempting registration with APNS');
          success = (token) => {
            this.trigger('_registration', {
              network: 'apns',
              token: token
            });
          };
          pushNotification.register(success, reject, {
            badge: true,
            sound: true,
            alert: true,
            ecb: "onNotificationAPNS"
          });
          break;
        default:
          reject(new Error('Invalid platform'));
      }
    });
  },

  /**
    Unregisters this specific device for push notifications.

    Example:

    ```javascript
    // app/routes/application.js
    import Em from 'ember';

    export default Em.Route.extend({
      actions: {
        unsubscribeNotifications: function(){
          this.notifications.unregister().then(function(){
            Em.Logger.info('Successfully unregistered the device');
          }).catch(function(error){
            Em.Logger.error('Something went wrong while unregistering', error);
          });
        }
      }
    });
    ```

    @method unregister
    @return {RSVP.Promise} A promise that will resolve when registration is successful.
  */
  unregister: function(){
    return new Em.RSVP.Promise((resolve, reject) => {
      const pushNotification = window.plugins.pushNotification;
      pushNotification.unregister(resolve, reject, {});
    });
  },

  /**
    Sets up callbacks expected by PushPlugin
    @method _subscribeForEcb
    @private
  */
  _subscribeForEcb: function(){
    window.onNotificationGCM = (e) => {
      switch (e.event) {
        case 'registered':
          if (e.regid.length > 0) {
            Em.Logger.info('Received regid %s from GCM', e.regid);
            this.trigger('_registration', {
              network: 'gcm',
              token: e.regid
            });
          }
          break;
        case 'message':
          // if this flag is set, this notification happened while we were in the foreground.
          // you might want to play a sound to get the user's attention, throw up a dialog, etc.
          if (e.foreground) {
            // on Android soundname is outside the payload.
            // On Amazon FireOS all custom attributes are contained within payload
            let soundfile = e.soundname || e.payload.sound;
            if (soundfile) {
              this.trigger('sound', '/android_asset/www/' + soundfile);
            }

            if (e.payload.message) {
              this.trigger('alert', e.payload.message);
            }

            if (e.payload.msgcnt) {
              this.trigger('badge', e.payload.msgcnt);
            }
          } else if (e.coldstart) {
            // App was launched completely
            // TODO: Handle events
          } else {
            // App was launched from a running background activity
            // TODO: Handle events
          }

          break;
        default:
          Em.Logger.error('Unable to handle %s event type from GCM.', e.event);
      }
    };

    window.onNotificationAPNS = (event) => {
      if (event.alert) {
        this.trigger('alert', event.alert);
      }

      if (event.sound) {
        this.trigger('sound', event.sound);
      }

      if (event.badge) {
        this.trigger('badge', event.badge);
      }
    };
  },

  /**
    Sets the badge count on the app icon if possible. Fails silently.
    @method _setBadgeCount
    @param {Number} badgeCount The current badge count.
    @private
  */
  _setBadgeCount: function(badgeCount){
    new Em.RSVP.Promise(function(resolve, reject){
      const pushNotification = window.plugins.pushNotification;
      pushNotification.setApplicationIconBadgeNumber(resolve, reject, badgeCount);
    }).catch(function(e){
      Em.Logger.error("Failed to set badge. %s", e.message);
    });
  },

  /**
    Plays a sound on the device. Fails silently.
    @method _playSound
    @param {String} filename Filename for the sound to play.
    @private
  */
  _playSound: function(filename){
    if (!window.Media) {
      Em.Logger.warn("Unable to play sound. Media object not available.");
      return;
    }
    let sound = new window.Media(filename);
    sound.play();
  }
});
